library/core/src/ptr/non_null.rs

```diff
@@ -278,6 +352,114 @@ impl<T> NonNull<[T]> {
+    ///
+    /// [valid]: crate::ptr#safety
+    /// [`NonNull::dangling()`]: NonNull::dangling
+    /// [`pointer::offset`]: ../../std/primitive.pointer.html#method.offset
+    #[inline]
+    #[unstable(feature = "ptr_as_uninit", issue = "75402")]
+    pub unsafe fn as_uninit_slice(&self) -> &[MaybeUninit<T>] {
+        // SAFETY: the caller must uphold the safety contract for `as_uninit_slice`.
+        unsafe { slice::from_raw_parts(self.cast().as_ptr(), self.len()) }
+    }
+
+    /// Returns a unique reference to a slice of possibly uninitialized values. In contrast to
+    /// [`as_mut`], this does not require that the value has to be initialized.
+    ///
+    /// For the shared counterpart see [`as_uninit_slice`].
+    ///
+    /// [`as_mut`]: #method.as_mut
+    /// [`as_uninit_slice`]: #method.as_uninit_slice
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that all of the following is true:
+    ///
+    /// * The pointer must be [valid] for reads for `ptr.len() * mem::size_of::<T>()` many bytes,
+    ///   and it must be properly aligned. This means in particular:
+    ///
+    ///     * The entire memory range of this slice must be contained within a single allocated object!
+    ///       Slices can never span across multiple allocated objects.
+    ///
+    ///     * The pointer must be aligned even for zero-length slices. One
+    ///       reason for this is that enum layout optimizations may rely on references
+    ///       (including slices of any length) being aligned and non-null to distinguish
+    ///       them from other data. You can obtain a pointer that is usable as `data`
+    ///       for zero-length slices using [`NonNull::dangling()`].
+    ///
+    /// * The total size `ptr.len() * mem::size_of::<T>()` of the slice must be no larger than `isize::MAX`.
+    ///   See the safety documentation of [`pointer::offset`].
+    ///
+    /// * You must enforce Rust's aliasing rules, since the returned lifetime `'a` is
+    ///   arbitrarily chosen and does not necessarily reflect the actual lifetime of the data.
+    ///   In particular, for the duration of this lifetime, the memory the pointer points to must
+    ///   not get accessed (read or written) through any other pointer.
+    ///
+    /// This applies even if the result of this method is unused!
+    ///
+    /// See also [`slice::from_raw_parts_mut`][].
+    ///
+    /// [valid]: crate::ptr#safety
+    /// [`NonNull::dangling()`]: NonNull::dangling
+    /// [`pointer::offset`]: ../../std/primitive.pointer.html#method.offset
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// #![feature(allocator_api, ptr_as_uninit)]
+    ///
+    /// use std::alloc::{AllocRef, Layout, Global};
+    /// use std::mem::MaybeUninit;
+    /// use std::ptr::NonNull;
+    ///
+    /// let memory: NonNull<[u8]> = Global.alloc(Layout::new::<[u8; 32]>())?;
+    /// // This is safe as `memory` is a valid pointer to `[u8]`
```

Comment:
It's not initialized though, so it's not really valid?

Reply:
From `core::ptr#Safety`:

> For a pointer to be valid, it is necessary, but not always sufficient, that the pointer be dereferenceable: the memory range of the given size starting at the pointer must all be within the bounds of a single allocated object.

The safety constraint from `AllocRef` requires this:
> Memory blocks returned from an allocator must point to valid memory.

We have to elaborate on this in the allocator-wg, but this is another topic. Basically it is required for all allocators, that a valid pointer is returned. For this, there are currently two issues open: https://github.com/rust-lang/wg-allocators/issues/65 https://github.com/rust-lang/wg-allocators/issues/46. The former wants to refine the quoted phrase from `AllocRef`.

To sum up: This is not initialized, but valid. As long as we don't return `&[T]` this should be fine.

, that it's dereferencable. An allocator returns a well-aligned pointer to an allocated object. The pointer does not necessarily need to point to initialized memory for this method to be called. I think this is the whole point of `MaybeUninit`.

Reply:
That is *valid for a given access kind and size*. But what your text here says sounds more like "valid for a given type", which -- for me at least -- is something very different.

Reply:
I don't think I quite follow. Which validity invariant am I missing here?

The whole reason to use an allocator (and let's assume that `Global` is implemented correctly here) is to get back a memory region that we can write to. Such a memory area may contain anything, uninitialized, but has a fixed size of at least `NonNull::len()`. For me this is just like returning a slice of `MaybeUninit<u8>` with a size of `len()`. Since `alloc` returns `NonNull<[u8]>`, it is guaranteed that no padding exists between `ptr[0]` and `ptr[1]` as [`u8` has no padding bits and its size is exactly one byte](https://rust-lang.github.io/unsafe-code-guidelines/layout/scalars.html#fixed-width-integer-types). So why is `&[MaybeUninit<u8>]` not a valid representation?

It may well be that I overlook or misunderstand something, but at the moment I cannot see the reason.

Reply:
> Which validity invariant am I missing here?

`u8` must be initialized. So allocators do not return valid `u8`.

Your safety docs get it right, it says

> The pointer must be valid for reads (and writes) for `ptr.len() * mem::size_of::<T>()` many bytes

So the code's safety comment should match those docs. But they do not.
The correct thing to say would be something like "`memory` is valid for read and writes for `memory.len()` many bytes".

Reply:
Oh, I'm so glad. I was a little shocked :D

Sure, the comment is very poor worded and "valid" isn't the right term here. I think your suggestion makes it more clear:

```suggestion
    /// // This is safe as `memory` is a valid for reads and writes for `memory.len()` many bytes.
    /// // Note, that calling `memory.as_mut()` is not allowed here as the content may be uninitialized.
```

Reply:
"is a valid" -> "is valid"

Also I don't think there should be a comma after "Note".